java/dagger/internal/codegen/validation/InjectValidator.java

```diff
@@ -332,6 +340,15 @@ private void checkInjectIntoPrivateClass(
     }
   }
 
+  private void checkInjectIntoKotlinObject(
+      TypeElement element, ValidationReport.Builder<TypeElement> builder) {
+    if (kotlinMetadataUtil.isObjectClass(element)) {
```

Comment:
We should also check if its a Companion Object:
`kotlinMetadataUtil.isObjectClass(element) || kotlinMetadataUtil.isCompanionObjectClass(element)`